### PR TITLE
Allow tracking images as datauri

### PR DESCRIPTION
**Pull request by Ndpnt**
Re-implementation of #41

### CODE REVIEW

File: services/GooglePlayStore.filters.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import DatauriParser from 'datauri/parser.js';
```
Comment (MattiSG) [inline services/GooglePlayStore.filters.js:1]:
```suggestion
import DataURIParser from 'datauri/parser.js';
```

File: services/GooglePlayStore.filters.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import fetch from '../src/fetcher/index.js';
+
+const dataUriParser = new DatauriParser();
```
Comment (MattiSG) [inline services/GooglePlayStore.filters.js:4]:
```suggestion
const dataURI = new DataURIParser();
```

File: services/GooglePlayStore.filters.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+    const imageUrl = new URL(src, baseUrl).href; // Ensure url is absolute
+
+    const blob = await fetch(imageUrl, { returnBlob: true });
+
+    const extension = `.${blob.type.split('/')[1].split('+')[0]}`;
+    const arrayBuffer = await blob.arrayBuffer();
+
+    images[index].src = dataUriParser.format(extension, arrayBuffer).content;
+  }));
```
Comment (MattiSG) [inline services/GooglePlayStore.filters.js:22]: No promise is returned 🤔

File: services/GooglePlayStore.filters.js [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  return Promise.all(images.map(async ({ src }, index) => {
+    if (src.startsWith('data:')) {
+      return Promise.resolve(); // Already a data-URI, skip.
+    }
+
+    const imageUrl = new URL(src, baseUrl).href; // Ensure url is absolute
+
+    const blob = await fetch(imageUrl, { returnBlob: true });
+
+    const extension = `.${blob.type.split('/')[1].split('+')[0]}`;
```
Comment (MattiSG) [inline services/GooglePlayStore.filters.js:18]: > I would feel more comfortable relying on a MIME-type handling library such as [`mime`](https://www.npmjs.com/package/mime) 🙂

File: services/GooglePlayStore.json [hunk 5]
```diff
@@ -11,6 +11,9 @@
     },
     "Brand Guidelines": {
       "fetch": "https://developer.android.com/distribute/marketing-tools/brand-guidelines",
+      "filter": [
+        "convertImagesToDataURI"
```
Comment (MattiSG) [inline services/GooglePlayStore.json:15]:
```suggestion
        "storeImagesInFile"
```
or maybe
```suggestion
        "downloadImages"
```

File: src/fetcher/index.js [hunk 6]
```diff
@@ -4,7 +4,7 @@ import HttpsProxyAgent from 'https-proxy-agent';
 
 const LANGUAGE = 'en';
 
-export default async function fetch(url) {
+export default async function fetch(url, { returnBlob = false } = {}) {
```
Comment (MattiSG) [inline src/fetcher/index.js:7]:
```suggestion
export default async function fetch(url, { asRawData = false } = {}) {
```

File: src/fetcher/index.test.js [hunk 7]
```diff
@@ -10,22 +10,32 @@ const __dirname = path.dirname(new URL(import.meta.url).pathname);
...
   describe('#fetch', () => {
     it('returns the web page content of the given URL', async () => {
       const result = await fetch('https://www.facebook.com/terms.php');
       expect(result).to.be.equal(facebookTermsHTML);
     });
 
+    context('With returns as blob option', () => {
+      it('returns the blob of the given URL', async () => {
+        const result = await fetch('https://www.facebook.com/terms.php', { returnBlob: true });
+        expect(result.constructor.name).to.equal('Blob');
```
Comment (MattiSG) [inline src/fetcher/index.test.js:35]:
```suggestion
        expect(result).to.be.an.instanceOf(Blob);
```

File: src/filter/index.js [hunk 8]
```diff
@@ -10,12 +10,26 @@ turndownService.use(turndownPluginGithubFlavouredMarkdown.gfm);
...
+    filter: serviceSpecificFilters = [],
+  } = documentDeclaration;
+
   const { document: webPageDOM } = new JSDOM(content).window;
 
-  serviceSpecificFilters.forEach(filterName => {
-    filterFunctions[filterName](webPageDOM); // filters work in place
-  });
+  for (const filterName of serviceSpecificFilters) {
+    const isAsyncFilter = filterFunctions[filterName].constructor.name === 'AsyncFunction';
```
Comment (MattiSG) [inline src/filter/index.js:24]: > Extract a function `isAsync` that checks if its parameter is an `instanceof` a cached `(async () => {}).constructor` 🙂  🌱

**Review by MattiSG — CHANGES_REQUESTED** [review]
Data URI is generated and stored! 👍
…but the resulting Markdown doesn't show any image 🙁

File: services/GooglePlayStore.filters.js (hunk 3 above)
Comment (Ndpnt) [reply services/GooglePlayStore.filters.js:22]: Not necessary. As function is declared `async`, the return value is automatically wrapped in a promise.

File: src/fetcher/index.test.js (hunk 7 above)
Comment (Ndpnt) [reply src/fetcher/index.test.js:35]: `Blob` is not exposed in Node.js

File: services/GooglePlayStore.filters.js (hunk 4 above)
Comment (MattiSG) [reply services/GooglePlayStore.filters.js:18]: Will be done in #109.

**Review by MattiSG — APPROVED** [review]
Tested locally, works!

# :clap: 🚀

**Pull request merged**